### PR TITLE
v1alpha2 Workflow Controller

### PR DESCRIPTION
**Pull request by chrisdoherty4**
Introduce basic functionality for a v1alpha2 Workflow Controller. This implementation is independent of the existing controller as we won't support the v1alpha1 API when we move to v1alpha2. The first commit moves deprecated code into /internal/deprecated and isn't that interesting: the work is mostly in the [second commit](https://github.com/tinkerbell/tink/pull/874/commits/eb5eb09b37ab79ea4c346422712d9e3929beacc0).

The implementation renders the template actions onto the Workflow only which will aid in E2E tests with Tink Server. Logic such as remediating Workflow states is deferred to a later PR.

### CODE REVIEW

File: Makefile
Comment (jacobweinstock) [inline Makefile:52]: not too excited we have to use `tink-controller-v1alpha2` :(

File: api/v1alpha2/conditions.go
Comment (jacobweinstock) [inline api/v1alpha2/conditions.go:33]: I know you love pointers ;) so why move away from one here?

File: cmd/tink-controller-v1alpha2/main.go
Comment (jacobweinstock) [inline cmd/tink-controller-v1alpha2/main.go:1]: is v1alpha2 so different from a cmd/cli perspective that we have to break it out entirely? Could a flag be used in the existing tink-controller?

File: cmd/tink-controller-v1alpha2/Dockerfile
Comment (jacobweinstock) [inline cmd/tink-controller-v1alpha2/Dockerfile:1]: any reason not to use `scratch`?

File: cmd/tink-controller-v1alpha2/main.go
Comment (jacobweinstock) [inline cmd/tink-controller-v1alpha2/main.go:140]: do we need a default location for a config file? does this provide much value, you think?

File: internal/workflow/internal/reconcile_test.go
Comment (jacobweinstock) [inline internal/workflow/internal/reconcile_test.go:1]: im not fond of `internal_test` files that use dot notation import for the package it's testing. I don't see the value. Not anything to hold up the PR, just wanted to voice it. : )

File: internal/workflow/reconciler.go
Comment (jacobweinstock) [inline internal/workflow/reconciler.go:36]: are we gaining any value by having named return values? I don't see them being used, feels unnecessary.

File: internal/workflow/reconciler.go
Comment (jacobweinstock) [inline internal/workflow/reconciler.go:59]: Why always attempt a patch?

**Review by jacobweinstock — COMMENTED** [review]
Thanks for the work on this, @chrisdoherty4. Left some comments for discussion.

File: cmd/tink-controller-v1alpha2/main.go
Comment (chrisdoherty4) [reply cmd/tink-controller-v1alpha2/main.go:1]: I'm open to using a single binary. This temporary extra binary just felt simpler to me. There won't be a transition period where both APIs are supported so eventually `tink-controller-v1alpha2` will become `tink-controller`.

File: cmd/tink-controller-v1alpha2/Dockerfile
Comment (chrisdoherty4) [reply cmd/tink-controller-v1alpha2/Dockerfile:1]: I don't know. The goal of the PR is to get something functional so I've not experimented.

File: cmd/tink-controller-v1alpha2/main.go
Comment (chrisdoherty4) [reply cmd/tink-controller-v1alpha2/main.go:140]: Probably not, copy-n-paste from the existing tink-controller. If we maintain the temporary v1alpha2 binary, I'll reduce this down to bare minimum so we don't end up with cruft down the road.

File: internal/workflow/internal/reconcile_test.go
Comment (chrisdoherty4) [reply internal/workflow/internal/reconcile_test.go:1]: Do you mean specifically for `internal` packages, or do you mean generally the `_test` idiom with dot import?

**Review by jacobweinstock — APPROVED** [review]

**Comment by jacobweinstock** [comment]
Hey @chrisdoherty4 , feel free to merge as is. We can address these before the official removal of v1alpha1.

File: internal/workflow/reconciler.go
Comment (chrisdoherty4) [reply internal/workflow/reconciler.go:36]: The named returns are included for the defer to work because we change the return variable if patching fails. I don't think the code needs to work this way, and honestly I'm not fond of the defer approach - defering the patch is a common method for always patching.

File: internal/workflow/reconciler.go
Comment (chrisdoherty4) [reply internal/workflow/reconciler.go:59]: The rationale is that you do as much as possible to bring the resource to the desired state and you shouldn't update the `.Status` state if nothing changed so always patching ensures we always try to persist the updates.

File: internal/workflow/reconciler.go
Comment (chrisdoherty4) [reply internal/workflow/reconciler.go:36]: Well, that's a good pont: `_`.

**Review by jacobweinstock — APPROVED** [review]

**Pull request merged**
